osdk/src/commands/profile.rs

```diff
@@ -55,31 +62,63 @@ fn do_collect_stack_traces(args: &ProfileArgs) {
 
     println!("Profiling \"{}\" at \"{}\".", file_path.display(), remote);
+    // Use GDB to halt the remote, get stack traces, and resume
+    let mut gdb_process = {
+        let file_cmd = format!("file {}", file_path.display());
+        let target_cmd = format!("target remote {}", remote);
+        let backtrace_cmd_seq = vec![
+            "-ex",
+            "thread apply all bt -frame-arguments presence -frame-info short-location",
+            "-ex",
+            "continue",
+        ];
+
+        let mut gdb_args = vec![
+            "-batch",
+            "-ex",
+            "set pagination 0",
+            "-ex",
+            &file_cmd,
+            "-ex",
+            &target_cmd,
+        ];
+        gdb_args.append(&mut vec![backtrace_cmd_seq; *samples].concat());
+
+        Command::new("gdb")
+            .args(gdb_args)
+            .stdout(Stdio::piped())
+            .spawn()
+            .expect("Failed to execute gdb")
+    };
+
+    let mut gdb_output = String::new();
+    let gdb_stdout = gdb_process.stdout.take().unwrap();
+    let mut gdb_output_reader = std::io::BufReader::new(gdb_stdout);
     use indicatif::{ProgressIterator, ProgressStyle};
     let style = ProgressStyle::default_bar().progress_chars("#>-");
     for _ in (0..*samples).progress_with_style(style) {
-        // Use GDB to halt the remote, get stack traces, and resume
-        let output = Command::new("gdb")
-            .args([
-                "-batch",
-                "-ex",
-                "set pagination 0",
-                "-ex",
-                &format!("file {}", file_path.display()),
-                "-ex",
-                &format!("target remote {}", remote),
-                "-ex",
-                "thread apply all bt -frame-arguments presence -frame-info short-location",
-            ])
-            .output()
-            .expect("Failed to execute gdb");
-
-        for line in String::from_utf8_lossy(&output.stdout).lines() {
-            profile_buffer.append_raw_line(line);
+        let mut buf = String::new();
+        loop {
+            let _ = gdb_output_reader.read_line(&mut buf);
+            gdb_output.push_str(&buf);
+            if buf.ends_with("0x0000000000000000 in ?? ()\n") {
+                break;
+            }
+            buf.clear();
         }
+        let _ = Command::new("kill")
+            .args(["-INT", &format!("{}", gdb_process.id())])
+            .spawn();
+        thread::sleep(time::Duration::from_secs_f64(*interval));
+    }
+    // Sleep for a while until gdb process fully exits, otherwise we kill it explicitly.
+    thread::sleep(time::Duration::from_secs_f64(5.0));
+    if let Ok(None) = gdb_process.try_wait() {
+        let _ = gdb_process.kill();
+    }
 
-        // Sleep between samples
-        std::thread::sleep(std::time::Duration::from_secs_f64(*interval));
+    for line in gdb_output.lines() {
+        profile_buffer.append_raw_line(line);
```

Comment:
How about adding the lines to the `profile_buffer` in the sample loop at line 113 after sending INT? Keeping the raw output in the vector seems less space-efficient then the parsed format in `profile_buffer`. And it makes parsing and INTing parallel.